File: python/m_auth/l3c/agent/graph.py
from langchain_core.messages import AIMessage
from langchain_core.runnables import RunnableConfig
from langgraph.graph import END, START, MessagesState, StateGraph


def chat(state: MessagesState, config: RunnableConfig) -> dict:
    user = config["configurable"].get("langgraph_auth_user", {})
    identity = user.get("identity", "unknown")
    last = state["messages"][-1].content
    return {"messages": [AIMessage(content=f"[{identity}] {last}")]}


graph = StateGraph(MessagesState)
graph.add_node("chat", chat)
graph.add_edge(START, "chat")
graph.add_edge("chat", END)
graph = graph.compile()

File: typescript/m5/l3a/agent/graph.ts
import { END, START, MessagesAnnotation, StateGraph, type LangGraphRunnableConfig } from "@langchain/langgraph";
import { AIMessage } from "@langchain/core/messages";

function chat(
  state: typeof MessagesAnnotation.State,
  config: LangGraphRunnableConfig,
): typeof MessagesAnnotation.Update {
  const user = (config.configurable?.langgraph_auth_user ?? {}) as Record<string, unknown>;
  const identity = (user.identity as string) ?? "unknown";
  const last = state.messages[state.messages.length - 1];
  const content = typeof last.content === "string" ? last.content : "";
  return { messages: [new AIMessage(`[${identity}] ${content}`)] };
}

export const graph = new StateGraph(MessagesAnnotation)
  .addNode("chat", chat)
  .addEdge(START, "chat")
  .addEdge("chat", END)
  .compile();
